kytos/core/interface.py

```diff
@@ -451,28 +389,62 @@ def add_tags(self, tags: list[int], tag_type: str = 'vlan') -> bool:
     def make_tags_available(
         self,
         controller,
-        tags: Union[int, list[int]],
+        tags: Union[int, list[int], list[list[int]]],
         tag_type: str = 'vlan',
         use_lock: bool = True,
-    ) -> bool:
+    ) -> list[list[int]]:
         """Add a specific tag in available_tags.
 
         Args:
             controller: Kytos controller
             tags: List of tags, there should be 2 items in the list
             tag_type: TAG type value
             use_lock: Boolean to whether use a lock or not
+
+        Return:
+            conflict: Return any values that were not added.
         """
+        if not tags:
+            return
         if isinstance(tags, int):
             tags = [tags] * 2
+        if isinstance(tags[0], int) and tags[0] != tags[1]:
+            tags = [tags[0], tags[1]]
         if use_lock:
             with self._tag_lock:
-                result = self.add_tags(tags, tag_type)
+                try:
+                    conflict = self._make_tags_available(tags, tag_type)
+                except KytosTagsNotInTagRanges as err:
+                    raise err
         else:
+            try:
+                conflict = self._make_tags_available(tags, tag_type)
+            except KytosTagsNotInTagRanges as err:
+                raise err
+        self._notify_interface_tags(controller)
+        return conflict
+
+    def _make_tags_available(
+        self,
+        tags: Union[list[int], list[list[int]]],
+        tag_type: str,
+    ) -> list[list[int]]:
+        """Manage available_tags adittion changes"""
```

Comment:
The returned conflict is behaving differently when additional `list[int]` are passed:

- Let's say a NApp used `[1, 10]`, and then let's say it tries to make available `[1, 12]`

```
kytos $> intf.use_tags(controller, [1, 10])                                                                                                                                              

kytos $> intf.make_tags_available(controller, [1, 12])                                                                                                                                   
Out[15]: [[1, 12]]

kytos $> intf.available_tags                                                                                                                                                             
Out[16]: {'vlan': [[11, 3798], [3800, 4095]]}

kytos $> intf.make_tags_available(controller, [[1, 12]])                                                                                                                                 
Out[17]: [[11, 12]]

kytos $> intf.available_tags                                                                                                                                                             
Out[18]: {'vlan': [[1, 3798], [3800, 4095]]}
```

It'd be great to standardize this, otherwise clients might find it confusing, food for thought, while still maintaining bisect, let me know what you want to propose here. At the end of the day a `list[int]` is a subcase of `list[list[int]]`, although initially we weren't returning the conflict part and it was all or nothing.

Reply:
I made a mistake here, so the idea was to use `add_tags()` when a single value was going to be added for example `12` or `[12,12]`. For multiple values `range_addition()` is used, `[1,12]` or `[[1,2], [5,6]]`. The `[1,12]` should have been inserted in a list `[[1,12]]` and be processed by `range_addition()`.
An old option I had was to use `add_tags()` in all cases but if one range could not be added then `range_addition()` would try to do it. This was based on the idea that everything should be working fine and `range_addition()` would only be called when an error most likely happened when working with other VLANs which is very unlikely.

Reply:
I see. That also implies not using a `bisect` for a case `[a, b]` where a != b that we used to, although an insertion after bisecting was also linear, so in the end both time complexity I'd expect to be relatively similar, with that said, I'll encourage you to also benchmark if there's still uncertainty. Once e2e tests and if local stress tests are all set, let me know and I'll do a final review. Other than that, it's shaping up really well. Thanks, Aldo.

Reply:
The `bisect` would be always used except when an error was detected. When an error is detected then `range_adittion()` takes its place.
Also, It looks like the current implementation is working correctly, [comment](https://github.com/kytos-ng/mef_eline/pull/396#issuecomment-1803357748). I tested it with concurrent evc creations. It takes time though.

Reply:
@Alopalao, on `interface.use_tags` yes, but on `interface.make_tags_available` it'll keep using the `range_addition` and `range_difference` when there's just one tag range if a != b, `[[a, b]]`, for example, notice the temporary prints that I added to illustrate this. Notice beleow it only executed `add_tags` when it was passed `[12, 12]`, but not `[[12, 12]]`:

```
diff --git a/kytos/core/interface.py b/kytos/core/interface.py
index 5953bec..5b764b0 100644
--- a/kytos/core/interface.py
+++ b/kytos/core/interface.py
@@ -349,6 +349,7 @@ class Interface(GenericEntity):  # pylint: disable=too-many-instance-attributes
         available_tag e.g [[7, 10], [20, 30], [78, 92], [100, 109], [189, 200]]
         tags examples are in each if statement.
         """
+        LOG.info(f"running add_tags for {tags}, tag_type: {tag_type}")
         if not tags[0] or not tags[1]:
             return False
 
@@ -441,6 +442,7 @@ class Interface(GenericEntity):  # pylint: disable=too-many-instance-attributes
             KytosTagsNotInTagRanges from add_tags()
         """
         if isinstance(tags[0], list):
+            LOG.info(f"computing range_difference and later range_addition for {tags}, {tag_type}")
             diff = range_difference(tags, self.tag_ranges[tag_type])
             if diff:
                 raise KytosTagsNotInTagRanges(diff, self._id)
```

```
kytos $> intf.make_tags_available(controller, [1, 12])                                                                                                                                    
Out[4]: [[11, 12]]

kytos $> 2023-11-13 15:07:47,556 - INFO [kytos.core.interface] [interface.py:445:_make_tags_available] (ThreadPoolExecutor-1_0) computing range_difference and later range_addition for [[
1, 12]], vlan
kytos $>                                                                                                                                                                                  

kytos $> intf.make_tags_available(controller, [[1, 12]])                                                                                                                                  
Out[5]: [[1, 12]]

kytos $> 2023-11-13 15:08:05,706 - INFO [kytos.core.interface] [interface.py:445:_make_tags_available] (ThreadPoolExecutor-1_0) computing range_difference and later range_addition for [[
1, 12]], vlan
kytos $>                                                                                                                                                                                  

kytos $> intf.make_tags_available(controller, [1, 12])                                                                                                                                    
Out[6]: [[1, 12]]

kytos $> 2023-11-13 15:08:13,643 - INFO [kytos.core.interface] [interface.py:445:_make_tags_available] (ThreadPoolExecutor-1_0) computing range_difference and later range_addition for [[
1, 12]], vlan
kytos $>                                                                                                                                                                                  

kytos $> intf.make_tags_available(controller, [12, 12])                                                                                                                                   
Out[7]: [[12, 12]]

kytos $> 2023-11-13 15:09:03,739 - INFO [kytos.core.interface] [interface.py:352:add_tags] (ThreadPoolExecutor-1_0) running add_tags for [12, 12], tag_type: vlan                         
kytos $>                                                                                                                                                                                  

kytos $> intf.make_tags_available(controller, [[12, 12]])                                                                                                                                 
Out[8]: [[12, 12]]

kytos $> 2023-11-13 15:09:11,701 - INFO [kytos.core.interface] [interface.py:445:_make_tags_available] (ThreadPoolExecutor-1_0) computing range_difference and later range_addition for [[
12, 12]], vlan
kytos $>                                                                                                                                                                                  
```

That's why it's good to benchmark and also confirm, if there's a way to try to use `add_tags` when just a single tag range is used, that'd be great, maybe it could also try to return a conflict list? If you could try to assess this. But, again, this is me assuming that on average the bisect version would perform better, despite an insertion in certain cases. Good thing under normal circumstances, usually UNI range ops won't be that high frequency, and links make a `int` or `[int_a, int_b]` tag available where int_a == int_b will still use the bisect. Thanks.

Reply:
Yeah, I think the process is staying as it is. I made some ploting with `range_addition()` with `matplotlib.pyplot` library:

1. No conflicts, 1024 ranges:
```
available = [[1, 2], [5, 6], ... ]
tags = [[3, 4], [7, 8], ... ]
```
![ploting_1](https://github.com/kytos-ng/kytos/assets/55767214/2d92fc69-feb1-4ecf-a00b-2017797abcd6)

2. Conflicts, 1024 ranges:
```
available = [[1, 2], [4, 6], ... ]
tags = [[2, 4], [6, 8], ... ]
conflicts = [[2, 2], [6, 6] ... ]
```
![ploting_2](https://github.com/kytos-ng/kytos/assets/55767214/c097a391-5383-466d-a3e1-90b3c6b96f6a)

3. Conflicts, 1024 available ranges and 2047 tag ranges: 
```
available = [[2, 2], [6, 6], [10, 10]... ]
tags = [[1, 1], [3, 3], [6, 6]... ]
conflicts = [[6, 6], [10, 10], ... ]
```
![ploting_3](https://github.com/kytos-ng/kytos/assets/55767214/53c65c7e-4679-4fbf-bf4a-c428f4fb1de0)

4. Conflicts, 8 available ranges and 7 tag ranges: 
```
available = [[3, 10], [20, 50], [60, 70], [80, 90], [100, 110], [112, 120], [130, 140], [150, 160]]
tags = [[1, 1], [3,3], [9, 22], [24, 55], [57, 62], [81, 101], [123, 128]]
conflicts = [[3, 3], [9, 10], [20, 22], [24, 50], [60, 62], [81, 90], [100, 101]]
```
![ploting_4](https://github.com/kytos-ng/kytos/assets/55767214/589a8d72-eadb-494c-97e2-f26aa4c1964b)

5. Conflicts, 3 available ranges and 1 tag ranges: 
```
available = [[3, 10], [20, 30], [40, 50]]
tags = [[5, 18]]
conflicts = [[5, 10]]
```
![ploting_5](https://github.com/kytos-ng/kytos/assets/55767214/72704b2a-58b7-4fc1-bd1a-b4739465d431)

Reply:
Right. Let's keep it this way. One of the worse-ish cases is taking a few milliseconds (equivalent ish to a DB W/R indexed operation), which isn't that great, not not too bad either all things considered, and it's very unlikely that this will become a bottleneck, especially since highly partitioned or reconfiguration of vlan ranges will typically tend to be not too frequent. If we ever need to optimize this we'll revisit the options. Nicely done.